generator/src/main/java/com/graphicsfuzz/generator/mutateapi/FactManager.java

```diff
@@ -0,0 +1,116 @@
+import com.graphicsfuzz.common.ast.stmt.BlockStmt;
+import com.graphicsfuzz.common.ast.stmt.DeclarationStmt;
+import com.graphicsfuzz.common.ast.stmt.ExprStmt;
+import com.graphicsfuzz.common.ast.stmt.ReturnStmt;
+import com.graphicsfuzz.common.ast.stmt.Stmt;
+import com.graphicsfuzz.common.ast.type.BasicType;
+import com.graphicsfuzz.common.ast.type.QualifiedType;
+import com.graphicsfuzz.common.ast.visitors.StandardVisitor;
+import com.graphicsfuzz.common.util.IRandom;
+import com.graphicsfuzz.common.util.PipelineInfo;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class FactManager {
+
+  private final FactManager prototype;
+
+  private final Map<VariableDeclInfo, VariableFact> variableFacts;
+  private final Map<FunctionPrototype, FunctionFact> functionFacts;
+
+  public FactManager(FactManager prototype) {
+    this.prototype = prototype;
+    variableFacts = new HashMap<>();
+    functionFacts = new HashMap<>();
+  }
+
+  public static Expr generateExpr(TranslationUnit tu,
+                                  PipelineInfo pipelineInfo,
+                                  FactManager factManager,
+                                  FunctionDefinition functionDefinition,
+                                  Stmt stmt,
+                                  Value value,
+                                  IRandom generator) {
+
+    String primitiveValueString = value.getData().get(0).get().toString();
+
+    switch (generator.nextInt(2)) {
+      // variableFact
+      case 0:
+        String variableName =
+            "_GLF_PRIMITIVE_VAR_" + value.getType().toString() + "_" + primitiveValueString.replace("."
+                , "_");
+        VariableDeclInfo declInfo =
+            new VariableDeclInfo(variableName,
+                null,
+                new Initializer(new FloatConstantExpr(primitiveValueString)));
+        VariablesDeclaration variablesDecl = new VariablesDeclaration(value.getType(), declInfo);
+        functionDefinition.getBody().addStmt(new DeclarationStmt(variablesDecl));
+
+        // Store variable fact in FactManager
+        factManager.getVariableFacts().put(declInfo, new VariableFact(variablesDecl, declInfo,
+            value));
+        return new VariableIdentifierExpr(variableName);
+
+
+      case 1:
+        // functionFact
+
+        String functionName =
+            "_GLF_COMPUTE_"
+                + value.getType().toString()
+                + "_" + primitiveValueString.replace(".", "_");
+        List<Declaration> declarations = new ArrayList<>(tu.clone().getTopLevelDeclarations());
+        FunctionPrototype functionPrototype = new FunctionPrototype(functionName,
+            new QualifiedType(value.getType(), Arrays.asList()));
+
+        declarations.add(0, new FunctionDefinition(functionPrototype,
+            new BlockStmt(
+                Arrays.asList(
+                    new ReturnStmt(
```

Comment:
Here, instead of returning a FloatConstantExpr, you should recursively call generateExpr.  But you should do so with a new FactManager that has each of the function parameters with known values as facts.

For example, if you are generating a function with this signature

_GLF_COMPUTE_42(float _GLF_PARAM_12, float _GLF_PARAM_UNKNOWN, float GLF_PARAM_64) {
   return ...
}

then you want to generate the ... with the facts that _GLF_PARAM_12 has the value 12, and _GLF_PARAM_64 has value 64.

The new fact manager should have the same global-scope facts as the current fact manager (but as it's generating a new function, it can't remember any local scope facts, e.g. about local variables).